pre_commit/editor.py

```diff
@@ -0,0 +1,167 @@
+import psutil
+import argparse
+import shlex
+import contextlib
+import concurrent.futures
+import subprocess
+from pathlib import Path
+from typing import Any, Callable, List, NoReturn, TypeVar
+
+from pre_commit.metrics import monitor
+from pre_commit import output, git
+
+T = TypeVar('T')
+
+# Keeping the file name the same as git's makes it more likely that editors will set the file type
+# correctly when opening it.
+COMMIT_MESSAGE_DRAFT_PATH = Path('.git/pre-commit/COMMIT_EDITMSG')
+COMMIT_MESSAGE_EXPIRED_DRAFT_PATH = Path('.git/pre-commit/COMMIT_EDITMSG_OLD')
+COMMIT_MESSAGE_HEADER = """
+# Please enter the commit message for your changes. Lines starting
+# with '#' will be ignored, and an empty message aborts the commit.
+#
+"""
+
+
+def should_run_concurrently(hook_stage: str) -> bool:
+    return hook_stage == 'commit' and _is_editor_script_configured() and _should_open_editor()
+
+def run_concurrently(fun: Callable[..., T], *args: Any) -> T:
+    # Allow user to enter commit message concurrently with running pre-commit hooks to lower
+    # wait times.
+    output.write_line('Waiting for your editor (pre-commit hooks running in background)...')
+
+    # Run git commands before starting hooks to avoid race conditions and git lock errors.
+    commit_message_template = _get_commit_message_template()
+
+    with contextlib.ExitStack() as paused_stdout_stack:
+        paused_stdout_stack.enter_context(output.paused_stdout())
+        def launch_editor() -> None:
+            _edit_commit_message(commit_message_template)
+            paused_stdout_stack.close()  # Resume terminal output as soon as the editor closes
+        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as ex:
+            ex.submit(launch_editor)
+            retval_future = ex.submit(fun, *args)
+            return retval_future.result()
+    raise RuntimeError('unreachable')
+
+def should_clean_draft(hook_stage: str) -> bool:
+    # We need to clean up the draft if one exists but we're not editing it, otherwise it's at risk
+    # of being committed without further user interaction.
+    return (hook_stage == 'commit'
+            and _is_editor_script_configured()
+            and not _should_open_editor()
+            and COMMIT_MESSAGE_DRAFT_PATH.exists()
+           )
+
+def clean_draft() -> None:
+    COMMIT_MESSAGE_DRAFT_PATH.rename(COMMIT_MESSAGE_EXPIRED_DRAFT_PATH)
```

Comment:
I... think this may run into issues on Windows if the destination file already exists. The `path.rename` docs say:

> On Unix, if target exists and is a file, it will be replaced silently if the user has permission.

Whereas the `path.replace` docs say:

> If target points to an existing file or directory, it will be unconditionally replaced.

But also does Git even work in non-Bash-like environments? Still, might be nice to either use `replace` or explicitly remove the destination first if it exists.

EDIT: looks like there's a *bunch* in here that assumes *nix-like system (which again is probably fine) so this is probably a-okay as is haha.

Reply:
I'll add an explicit check to make sure we're not on Windows before firing any of this stuff.